### PR TITLE
L0 tests for the PRCA task.

### PR DESCRIPTION
**Pull request by HSAR**

### CODE REVIEW

File: Tasks/PullRequestCodeAnalysis/PRCA/PrcaOrchestrator.ts [hunk 1]
```diff
@@ -94,25 +95,40 @@ export class PrcaOrchestrator {
...
         this.logger.LogDebug(`SonarQube report path: ${sqReportPath}`);
         if (sqReportPath === undefined || sqReportPath === null) {
-            throw new Error("Make sure a SonarQube enabled build task ran before this step.");
+            return Promise.reject('Make sure a SonarQube-enabled build task ran before this step.');
         }
 
         var allMessages:Message[] = this.sqReportProcessor.FetchCommentsFromReport(sqReportPath);
         var messagesToPost:Message[] = null;
-        return this.PrcaService.getModifiedFilesInPr()
+        return Promise.resolve()
```
Comment (HSAR) [inline Tasks/PullRequestCodeAnalysis/PRCA/PrcaOrchestrator.ts]: This looks odd, but I needed an empty promise to start the chain so I could accomplish what happens below.

File: Tasks/PullRequestCodeAnalysis/PRCA/PrcaOrchestrator.ts [hunk 2]
```diff
@@ -94,25 +95,40 @@ export class PrcaOrchestrator {
...
+            .then(() => {
+                return this.PrcaService.getModifiedFilesInPr()
+                    .catch((error) => {
```
Comment (HSAR) [inline Tasks/PullRequestCodeAnalysis/PRCA/PrcaOrchestrator.ts]: This embedded catch serves to catch rejections that originate from _getModifiedFilesInPr()_ only, much like a conventional catch statement. We can then give a specific error message without the possibility of erroneously capturing rejections from further upstream.

Any rejection that does pass down from upstream will fall out of this Orchestrator (by design) and be caught by the master catch at the end of the PRCA task.

File: Tasks/PullRequestCodeAnalysis/Tests/L0.ts [hunk 3]
```diff
@@ -746,4 +750,53 @@ describe('The PRCA', function () {
             });
         });
     });
+
+    describe('task', () => {
+
+        it('succeeds but skips on a non-PR build', (done:MochaDone) => {
+            this.timeout(1000);
+
+            let tp: string = path.join(__dirname, 'L0skipsNotPr.js');
```
Comment (HSAR) [inline Tasks/PullRequestCodeAnalysis/Tests/L0.ts]: This is the new test system in action: "L0skipsNotPr.ts" now runs and sets up the test run, serving much as "response.json" used to in the old test infrastructure with the key change that **getVariable is no longer functional** and manual assignments in process.env is required (examples are in other task tests).

File: Tasks/PullRequestCodeAnalysis/Tests/L0failsWithoutReportLocation.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+process.env['BUILD_SOURCEBRANCH'] = 'refs/pull/6/master'; //task-lib doesn't support getVariable yet, so we need these
+process.env['BUILD_REPOSITORY_ID'] = '00000000-0000-0000-0000-000000000000';
+process.env['ENDPOINT_AUTH_SYSTEMVSSCONNECTION'] = '{\"scheme\":\"OAuth\", \"parameters\": {\"AccessToken\": \"foobar\"}}';
+process.env['SYSTEM_TEAMFOUNDATIONCOLLECTIONURI'] = 'https://youraccount.visualstudio.com/DefaultCollection';
+
+tmr.setInput('messageLimit', '100');
+
+// provide answers for task mock
+let a: ma.TaskLibAnswers = <ma.TaskLibAnswers>{
```
Comment (HSAR) [inline Tasks/PullRequestCodeAnalysis/Tests/L0failsWithoutReportLocation.ts]: This object is mostly a one-for-one replacement for response.json except for the aforemention missing getVariable aspect.

File: Tasks/PullRequestCodeAnalysis/prcatask.ts [hunk 5]
```diff
@@ -10,38 +12,52 @@ import { TaskLibLogger } from './TaskLibLogger';
...
+    var pullRequestId: number = Number.parseInt(sourceBranch.replace('refs/pull/', ''));
+    if (isNaN(pullRequestId)) {
+        tl.debug(`Expected pull request ID to be a number. Attempted to parse: ${sourceBranch.replace('refs/pull/', '')}`);
         // Looks like: "Could not retrieve pull request ID from the server."
         tl.setResult(tl.TaskResult.Failed, tl.loc('Error_InvalidPullRequestId'));
     }
 
     var orchestrator:PrcaOrchestrator = PrcaOrchestrator.CreatePrcaOrchestrator(tlLogger, collectionUrl, token, repositoryId, pullRequestId);
-    orchestrator.postSonarQubeIssuesToPullRequest(process.env)
+    orchestrator.postSonarQubeIssuesToPullRequest(tl.getVariable('PRCA_REPORT_PATH'))
```
Comment (HSAR) [inline Tasks/PullRequestCodeAnalysis/prcatask.ts]: New task variable "PRCA_REPORT_PATH" should be assigned to by tasks supporting SonarQube analysis. Engineering work is already done in Maven, Gradle remains to do. Future pull requests will address this.

File: Tasks/PullRequestCodeAnalysis/PRCA/PrcaOrchestrator.ts [hunk 6]
```diff
@@ -94,25 +95,43 @@ export class PrcaOrchestrator {
     public postSonarQubeIssuesToPullRequest(sqReportPath: string): Promise<void> {
         this.logger.LogDebug(`SonarQube report path: ${sqReportPath}`);
         if (sqReportPath === undefined || sqReportPath === null) {
-            throw new Error("Make sure a SonarQube enabled build task ran before this step.");
+            return Promise.reject('Make sure a SonarQube-enabled build task ran before this step.');
```
Comment (bgavrilMS) [inline Tasks/PullRequestCodeAnalysis/PRCA/PrcaOrchestrator.ts]: loc

File: Tasks/PullRequestCodeAnalysis/prcatask.ts [hunk 7]
```diff
@@ -10,38 +12,52 @@ import { TaskLibLogger } from './TaskLibLogger';
...
     // Looks like: "Skipping pull request commenting - this build was not triggered by a pull request."
     console.log(tl.loc('Error_NotPullRequest'));
 } else {
-    var pullRequestId: number = Number(sourceBranch.replace('refs/pull/', ''));
+
+    var messageLimitInput:string = tl.getInput('messageLimit');
+    var messageLimit: number = Number(messageLimitInput);
     if (isNaN(messageLimit)) {
-        tl.debug(`Expected pull request ID to be a number. Actual: ${sourceBranch.replace('refs/pull/', '')}`);
+        // Looks like: "Expected message limit to be a number, but instead it was NOT_A_NUMBER"
```
Comment (bgavrilMS) [inline Tasks/PullRequestCodeAnalysis/prcatask.ts]: You also need to protect against a negative numbers and non-integers. http://stackoverflow.com/questions/10834796/validate-that-a-string-is-a-positive-integer

Comment (HSAR) [inline Tasks/PullRequestCodeAnalysis/prcatask.ts]: Good point. In progress.

File: Tasks/PullRequestCodeAnalysis/PRCA/PrcaOrchestrator.ts (hunk 6 above)
Comment (HSAR) [inline Tasks/PullRequestCodeAnalysis/PRCA/PrcaOrchestrator.ts]: In progress.

Comment (HSAR) [inline Tasks/PullRequestCodeAnalysis/PRCA/PrcaOrchestrator.ts]: Done.

File: Tasks/PullRequestCodeAnalysis/prcatask.ts (hunk 7 above)
Comment (HSAR) [inline Tasks/PullRequestCodeAnalysis/prcatask.ts]: Done. Thanks for the link!

**Pull request merged**